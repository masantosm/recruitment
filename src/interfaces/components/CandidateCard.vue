<template>
  <div
    class="w-candidate h-candidate flex flex-col justify-between bg-light-grey border border-border rounded-xl p-2 box-border mb-4"
  >
    <div class="flex items-stretch">
      <p class="w-10/12 font-bold truncate text-normal">{{ userName }}</p>
      <div class="w-2/12 flex justify-end text-normal reñative">
        <CardOptions :user-data="candidateData"></CardOptions>
      </div>
    </div>
    <p class="text-grey text-x-small">Añadido por ATS</p>
    <div class="flex justify-between">
      <div class="flex items-center">
        <img class="mr-1" :src="clock" />
        <span class="text-grey text-x-small">{{ formatDate }}</span>
      </div>
      <div class="mr-1 cursor-pointer">
        <button @click="setShowModal(true)">
          <img :src="edit" />
        </button>
      </div>
    </div>
  </div>
  <ModalForm
    :open="showModal"
    :candidate="currentUser"
    @close-modal="setShowModal(false)"
  ></ModalForm>
</template>

<script lang="ts">
import { defineComponent, computed, ref } from 'vue'
import CardOptions from '../components/CardOptions.vue'
import ModalForm from '../components/ModalForm.vue'
import buttonImg from './icons/icon-button.svg'
import clock from '../components/icons/icon-clock.svg'
import edit from '../components/icons/icon-edit.svg'
import type { CandidateDTO } from '@/application/dtos/CandidateDTO'

export default defineComponent({
  name: 'Candidate-card',
  props: {
    candidateData: {
      type: Object,
      required: true
    }
  },
  components: {
    CardOptions,
    ModalForm
  },
  setup(props) {
    const showModal = ref<boolean>(false)

    const userName = computed<string>(
      () => `${props.candidateData.firstName} ${props.candidateData.lastName}`
    )

    const currentUser = computed<CandidateDTO>(() => {
      return {
        firstName: props.candidateData.firstName,
        lastName: props.candidateData.lastName,
        statusId: props.candidateData.statusId,
        vacancyId: props.candidateData.vacancyId,
        candidateId: props.candidateData.id
      }
    })

    const formatDate = computed<string>(() =>
      new Date(props.candidateData.updatedAt).toLocaleDateString('es', {
        day: '2-digit',
        month: '2-digit',
        year: '2-digit'
      })
    )

    const setShowModal = (value: boolean) => {
      showModal.value = value
    }

    return {
      buttonImg,
      clock,
      edit,
      formatDate,
      userName,
      showModal,
      currentUser,
      setShowModal
    }
  }
})
</script>
<style scoped lang="less"></style>
