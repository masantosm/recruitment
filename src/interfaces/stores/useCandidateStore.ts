import { defineStore } from 'pinia'
import CandidateService from '../../application/services/CandidateService'
import type { CandidateDTO } from '../../application/dtos/CandidateDTO'
import type { Status } from '../../domain/models/Status'

interface CandidateState {
  candidates: CandidateDTO[]
  vacancyStatus: Status[]
  vacancyId: string
}

export const useCandidateStore = defineStore('candidate', {
  state: (): CandidateState => ({
    candidates: [],
    vacancyStatus: [],
    vacancyId: '106ec090-e5a5-45af-b82f-a3b5eda4117c'
  }),

  actions: {
    async loadCandidateStatus(vacancyId: string) {
      const status = await CandidateService.fetchCandidateStatus(vacancyId)
      this.vacancyStatus = status.data
    },
    async loadCandidatesVacancy(vacancyId: string) {
      const status = await CandidateService.fetchCandidates(vacancyId)
      this.candidates = status.data
    },
    async updateCandidateStatus(candidateId: string, candidate: CandidateDTO, vacancyId: string) {
      await CandidateService.updateCandidate(candidateId, candidate)
      this.loadCandidatesVacancy(vacancyId)
    },
    async createNewCandidate(candidate: CandidateDTO, vacancyId: string) {
      await CandidateService.createCandidate(candidate)
      this.loadCandidatesVacancy(vacancyId)
    }
  },

  getters: {
    getCandidateStatus: (state): Status[] => {
      return state.vacancyStatus
    },
    getCandidates: (state): CandidateDTO[] => {
      return state.candidates
    },
    getVacancyID: (state) => {
      return state.vacancyId
    }
  }
})
