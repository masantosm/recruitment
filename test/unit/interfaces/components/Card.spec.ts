import { mount, shallowMount } from '@vue/test-utils'
import { describe, it, expect, beforeEach, vi } from 'vitest'
import CardComponent from '../../../../src/interfaces/components/Card.vue'
import CandidateCard from '../../../../src/interfaces/components/CandidateCard.vue'
import { useCandidateStore } from '../../../../src/interfaces/stores/useCandidateStore'

vi.mock('@/interfaces/stores/useCandidateStore', () => ({
  useCandidateStore: vi.fn()
}))

describe('CardComponent.vue', () => {
  const config = {
    id: 1,
    name: 'Test Name',
    order: 1
  }

  let candidateStoreMock: any

  beforeEach(() => {
    candidateStoreMock = {
      getCandidates: [
        { id: 1, statusId: 1, name: 'Candidate 1' },
        { id: 2, statusId: 2, name: 'Candidate 2' }
      ]
    }
    // Mock store implementation for each test
    useCandidateStore.mockReturnValue(candidateStoreMock)
  })

  it('renders correctly with props', () => {
    const wrapper = shallowMount(CardComponent, {
      props: { config }
    })
    const card = wrapper.find('[data-test-id="card"]')
    expect(card.exists()).toBe(true)
    expect(wrapper.exists()).toBe(true)
    expect(wrapper.find('span').text()).toBe(config.name)
    expect(wrapper.find('img').attributes('src')).toBe(
      `src/interfaces/components/icons/icon-${config.order}.svg`
    )
  })

  it('computes the correct class and icon based on the config order', () => {
    const wrapper = mount(CardComponent, {
      props: { config }
    })
    const hr = wrapper.find('hr')
    expect(hr.classes()).toContain('card__1') // For order 1, it should have class `card__1`
    expect(wrapper.find('img').attributes('src')).toBe('src/interfaces/components/icons/icon-1.svg')
  })

  it('renders the CandidateCard component when candidates are present', () => {
    const wrapper = mount(CardComponent, {
      props: { config }
    })

    const candidateCards = wrapper.findAllComponents(CandidateCard)
    expect(candidateCards.length).toBe(1) // Only 1 candidate with statusId 1
    expect(candidateCards[0].props('candidateData')).toEqual({
      id: 1,
      statusId: 1,
      name: 'Candidate 1'
    })
  })

  it('does not render CandidateCard component if there are no matching candidates', () => {
    // Mock the store to return no matching candidates
    candidateStoreMock.getCandidates = [{ id: 3, statusId: 3, name: 'Candidate 3' }]
    const wrapper = mount(CardComponent, {
      props: { config }
    })

    const candidateCards = wrapper.findAllComponents(CandidateCard)
    expect(candidateCards.length).toBe(0) // No candidates with statusId 1
  })

  it('matches the snapshot', () => {
    const wrapper = shallowMount(CardComponent, {
      props: { config }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
