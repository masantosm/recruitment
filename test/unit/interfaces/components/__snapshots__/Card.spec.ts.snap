// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardComponent.vue > matches the snapshot 1`] = `
"<div data-v-ee190728="" class="w-card min-h-card p-4 box-border rounded-xl border border-border mr-5" data-test-id="card">
  <div data-v-ee190728="" class="w-candidate pb-5">
    <hr data-v-ee190728="" class="h-2 rounded-full border-0 card__1">
    <div data-v-ee190728="" class="flex items-end mt-4"><img data-v-ee190728="" src="src/interfaces/components/icons/icon-1.svg" class="w-button h-button"><span data-v-ee190728="" class="ml-2 text-base font-bold">Test Name</span></div>
  </div>
  <div data-v-ee190728="">
    <candidate-card-stub data-v-ee190728="" candidatedata="[object Object]"></candidate-card-stub>
  </div>
</div>"
`;
